### PR TITLE
Handle exceptions thrown in File.visititems() correctly

### LINKED ISSUE #1740: visititems behavior changed
I'm throwing an exception from a visititems callback function, in this case to implement a timeout. This is mostly useful for hdf5 files with many items, where listing them potentially takes a long time. This worked fine with previous h5py versions, with 3.0.0 instead of the thrown `Exception`, it causes a `SystemError`:

```
something <HDF5 dataset "something": shape (128, 128), type "<f8">
Traceback (most recent call last):
  File "h5py/h5o.pyx", line 302, in h5py.h5o.cb_obj_simple
  File "/home/alex/.virtualenvs/hdf5test/lib/python3.8/site-packages/h5py/_hl/group.py", line 601, in proxy
    return func(name, self[name])
  File "h5test.py", line 7, in throwing
    raise Exception("yeah")
Exception: yeah

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "h5test.py", line 11, in <module>
    f.visititems(throwing)
  File "/home/alex/.virtualenvs/hdf5test/lib/python3.8/site-packages/h5py/_hl/group.py", line 602, in visititems
    return h5o.visit(self.id, proxy)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
SystemError: <built-in function visit> returned a result with an error set
```

This can be reproduced using the following:

```python
import h5py
import numpy as np

if __name__ == "__main__":
    def throwing(name, obj):
        print(name, obj)
        raise Exception("yeah")

    with h5py.File("/tmp/test.h5", "w") as f:
        f['something'] = np.ones((128, 128))
        f.visititems(throwing)
```

With h5py 2.10.0, the `Exception` can be properly caught:

```
Traceback (most recent call last):
  File "h5test.py", line 11, in <module>
    f.visititems(throwing)
  File "/home/alex/.virtualenvs/hdf5test/lib/python3.8/site-packages/h5py/_hl/group.py", line 565, in visititems
    return h5o.visit(self.id, proxy)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 355, in h5py.h5o.visit
  File "h5py/defs.pyx", line 1641, in h5py.defs.H5Ovisit_by_name
  File "h5py/h5o.pyx", line 302, in h5py.h5o.cb_obj_simple
  File "/home/alex/.virtualenvs/hdf5test/lib/python3.8/site-packages/h5py/_hl/group.py", line 564, in proxy
    return func(name, self[name])
  File "h5test.py", line 7, in throwing
    raise Exception("yeah")
Exception: yeah
```

Is it possible to revert back to the previous behavior of bubbling up the original exception, or as an alternative, have native support for a timeout in `visititems`? Thanks!

 * Operating System:  Debian unstable
 * Where Python was acquired: system Python

```
Summary of the h5py configuration
---------------------------------

h5py    3.0.0
HDF5    1.12.0
Python  3.8.5 (default, Aug  2 2020, 15:09:07) 
[GCC 10.2.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.19.4
cython (built with) 0.29.21
numpy (built against) 1.17.5
HDF5 (built against) 1.12.0
```

### PR DESCRIPTION
**Pull request by uellue**
Closes #1740

Thx @sk1p and @takluyver!

### CODE REVIEW

File: docs/whatsnew/3.1.rst [hunk 1]
```diff
@@ -10,6 +10,7 @@ Bug fixes
   :meth:`.Group.create_virtual_dataset` (:issue:`1732`).
 * Fix writing data as a list to a dataset with a sub-array data type
   (:issue:`1735`).
+* Fix exception handling in :meth:`.File.visititems` (:issue:`1740`)
```
Comment (takluyver) [inline docs/whatsnew/3.1.rst:13]: 3.1 was already released - your changes will be in 3.2. To add a release note, make a new file in the `news/` folder by copying `TEMPLATE.rst`. The files in this directory will be pulled together when we do a release.

More about this in the contribution docs: https://docs.h5py.org/en/stable/contributing.html#write-a-release-note

**Comment by takluyver** [comment]
Other than the release note, this LGTM. :slightly_smiling_face:

**Comment by uellue** [comment]
@takluyver many thanks for the pleasant interaction and help. Contributing to h5py was enjoyable!

**Comment by takluyver** [comment]
No problem! :smiley:

**Pull request merged**